File: 12-19-2024/Merge_Intervals.java
import java.util.Arrays;

public class Merge_Intervals {
    public static int[][] merge(int[][] intervals) {
        int n = intervals.length;
        int[][] ans = new int[n][2];
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        int ind = 0;
        for(int i=1;i<n;i++) {
            if(intervals[i-1][1] >= intervals[i][0]) {
                intervals[i][0] = intervals[i-1][0];
                intervals[i][1] = Math.max(intervals[i][1], intervals[i-1][1]);
            } else {
                ans[ind++] = intervals[i-1];
            }
        }
        ans[ind] = intervals[n-1];

        return Arrays.copyOfRange(ans, 0, ind+1);
    }

    public static void main(String[] args) {
        int[][] arr = new int[][]{{1,4},{2,3}};
        int[][] ans = merge(arr);
        for(int[] a : ans)
            System.out.println(Arrays.toString(a));
    }
}

File: 2026-04-03/Merge_Intervals.py
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        new_intervals = [intervals[0]]

        for start, end in intervals:
            prev_end = new_intervals[-1][1]

            if start <= prev_end:
                new_intervals[-1][1] = max(end, prev_end)
            else:
                new_intervals.append([start, end])

        return new_intervals


sol = Solution()
print(sol.merge([[1,3],[2,6],[8,10],[15,18]]))
print(sol.merge([[1,9], [2,3],[4,5]]))
print(sol.merge([[4,7],[1,4]]))
